### PR TITLE
io_uring: add newest features from liburing, up from kernel 5.11

### PR DESCRIPTION
**Pull request by Jiboo**
Hi, this is an update on PR #13519, this time with usages and tests.

Added missing structs/enums/values related to latest io_uring updates (somewhat between kernel 5.11 and 6.1).
Added openat2 syscall for linux only (to re-use the structs for io_uring openat2).
Introduced some API breaks:
- removed IORING_REGISTER_* prefix for some IORING_REGISTER enums,
- added flags param to writev,
- added optional opt_file_index param to some functions, instead of having *_direct variants,

I'm still pretty new to zig, and welcome any comment you might have, as I'd like this to be merged before continuing my work on a side project in which I'd like to use latest io_uring features.

Thanks for your time,
JB.

### CODE REVIEW

**Comment by Jiboo** [comment]
Sidenote: missing tests for sendmsg_zc, as my kernel does not currently supports it (only in 6.1 and i'm still at 6.0).

**Comment by Jiboo** [comment]
Before accepting this: I'll try to find a way to run unit-tests on qemu for all kernel betweens 5.1 and 6.1, to find possible issues with unsupported opcodes/registers, etc.

**Comment by Jiboo** [comment]
Sorry about the multiple failed CI already.
Last attempt looked stuck on register-restrictions test, after looking there was indeed a bug (readv on wrong fd), sorry for missing that, not sure why the test do pass locally even with the bug. I've pushed a fix.

**Comment by matu3ba** [comment]
This looks like `"sync-cancel"` never finishes.

**Comment by Jiboo** [comment]
> This looks like `"sync-cancel"` never finishes.

Yep, probably blocked at the copy_cqe() call which waits for the read to complete as it's probably not canceled/interrupted, because the CI kernel is at 5.17 and doesn't support REGISTER_SYNC_CANCEL, although I wouldn have expected the register_sync_cancel() call to return error.NotSupported.

I don't repro locally (on kernel 6.0), but I'm still investigating ways to automate running io_uring tests on multiple kernels versions for the CI of zig, to avoid adding too much regressions with this MR, and possibly having a test framework for future updates on io_uring for zig. Although I didn't find anything which seemed easy to use, and my current lead is to compile a bunch of kernel versions, and run them through qemu's "microvm", with a minimal init that then runs the io_uring tests compiled as a static executable, but seems way overkill and a bit over my league ([source1](https://mergeboard.com/blog/2-qemu-microvm-docker/) [source2](https://github.com/firecracker-microvm/firecracker/blob/main/docs/rootfs-and-kernel-setup.md)).

**Comment by matu3ba** [comment]
> with a minimal init that then runs the io_uring tests compiled as a static executable, but seems way overkill and a bit over my league ([source1](https://mergeboard.com/blog/2-qemu-microvm-docker/) [source2](https://github.com/firecracker-microvm/firecracker/blob/main/docs/rootfs-and-kernel-setup.md)).

I think that we would want "get the config out of a Kernel to reproduce via rebuild" anyway (meaning somehow getting the runtime + compilation config + have a way to feed it into kconfig) for 1. abi checks and 2. to have less aweful ways of telling the type size.

Take as (de)motivational example this PR, where I tried to chase around what Kconfig might have been configured with: https://github.com/ziglang/zig/pull/13926#issuecomment-1369161380. This does not scale and is unnececssary discouraging for contributors.

Getting the runtime config is usually relative simple
```
get Kernel config
cat /proc/config.gz | gunzip > running.config
ls /boot/config-*
ls /usr/src/linux*
```
However, there exists no guide or automatism to my knowledge and I dont know if further configurations are necessary.

How could I build a Kernel from such a `.config` file and run it inside kvm to test things?

**Comment by Jiboo** [comment]
Sorry for the slow work on this, I have a working prototype for running the tests on specific linux kernels through qemu, that could be integrated with the CI:

[5.0.21] 1 passed; 49 skipped; 0 failed.
[5.1.21] 1 passed; 49 skipped; 0 failed.
[5.2.21] 1 passed; 49 skipped; 0 failed.
[5.3.18] 1 passed; 49 skipped; 0 failed.
[5.4.228] 10 passed; 40 skipped; 0 failed.
[5.5.19] 10 passed; 40 skipped; 0 failed.
[5.6.19] 24 passed; 26 skipped; 0 failed.
[5.7.19] 28 passed; 22 skipped; 0 failed.
[5.8.18] 28 passed; 22 skipped; 0 failed.
[5.9.16] 28 passed; 22 skipped; 0 failed.
[5.10.162] 34 passed; 16 skipped; 0 failed.
[5.11.22] 33 passed; 17 skipped; 0 failed.
[5.12.19] 33 passed; 17 skipped; 0 failed.
[5.13.19] 34 passed; 16 skipped; 0 failed.
[5.14.21] 34 passed; 16 skipped; 0 failed.
[5.15.86] 38 passed; 12 skipped; 0 failed.
[5.16.20] 38 passed; 12 skipped; 0 failed.
[5.17.15] 39 passed; 11 skipped; 0 failed.
[5.18.19] 41 passed; 9 skipped; 0 failed.
[5.19.17] 44 passed; 6 skipped; 0 failed.
[6.0.17] 48 passed; 2 skipped; 0 failed.
[6.1.3] 49 passed; 1 skipped; 0 failed.

Takes about 1min30 on my machine if all kernels are already built.
This could also be useful for testing other linux related parts of the std lib, dunno.

This PR should still be considered WIP, I gottat do a cleaning pass on what tests are skipped and why.

File: lib/std/os/linux.zig [hunk 1]
```diff
@@ -3745,7 +3669,63 @@ pub const IOSQE_BUFFER_SELECT = 1 << @enumToInt(IOSQE_BIT.BUFFER_SELECT);
...
+
+/// Only one task is allowed to submit requests
+/// Available since Linux 6.0
+pub const IORING_SETUP_SINGLE_ISSUER = 1 << 12;
+
+/// Defer running task work to get events.
+/// Rather than running bits of task work whenever the task transitions
+/// try to do it just before it is needed.
+/// Available since Linux 6.1
+pub const IORING_SETUP_DEFER_TASKRUN = 1 << 13;
```
Comment (daurnimator) [inline lib/std/os/linux.zig:3725]: Can these be grouped under a struct?
e.g. 
```zig
pub const IORING_SETUP = struct {
  IOPOLL = 1 << 0;

  SQPOLL = 1 << 1;
```

Similar for other flags

File: lib/std/os/linux.zig [hunk 2]
```diff
@@ -3776,20 +3761,43 @@ pub const IORING_OP = enum(u8) {
...
+    SEND_ZC,
+    // Available since Linux 6.1
+    SENDMSG_ZC,
 
     _,
 };
 
+// io_uring_sqe.uring_cmd_flags (rw_flags in the Zig struct)
+
+/// use registered buffer; pass thig flag along with setting sqe->buf_index.
```
Comment (daurnimator) [inline lib/std/os/linux.zig:3797]:
```suggestion
/// use registered buffer; pass this flag along with setting sqe->buf_index.
```

**Comment by andrewrk** [comment]
Closing abandoned draft.

**Pull request closed without merging**